### PR TITLE
feat: add debug tools

### PR DESCRIPTION
**Pull request by HaokunX-intel**
Place debug tools in tools/debug to help query the enclave agent
quickly, which is originally created by @zhiwei-intel-h. Only
allow to be used in the development env.

Signed-off-by: Xing, Haokun <haokun.xin@intel.com>

### CODE REVIEW

File: tools/debug/protocols/protos/agent.proto [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+syntax = "proto3";
+
+package grpc;
+
+import "oci.proto";
+import "github.com/kata-containers/agent/pkg/types/types.proto";
+import "google/protobuf/empty.proto";
+
+// unstable
+service AgentService {
```
Comment (haosanzi) [inline tools/debug/protocols/protos/agent.proto:16]: Agent encalve does not need these rpc interfaces serving kata agent.

File: docs/develop-guideline.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    - remote attestation: it provides security information for the remove 
+    - local attestation
+    - the crate [attestation-agent](https://github.com/confidential-containers/attestation-agent) supports a native attesation client
+- signature verification: 
+    - some images are signed by the pulisher or the store or both, it need to verify these signatures
+    - the crate [attestation-agent](https://github.com/confidential-containers/attestation-agent) supports a native signature client 
+
+# 2. Development Guide
+
+
```
Comment (haosanzi) [inline docs/develop-guideline.md:32]: please remove this empty line

File: docs/develop-env-install.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+- The directory `[path-to-install]/sgxsdk/lib64` contains dynamic libraries, run the `ldconfig -v -n [path-to-install]/sgxsdk/lib64` to update the library reference.
+- The directory `[path-to-install]/sgxsdk/include` contains the header files, if compiler cannot find these files, use `export C_INCLUDE_PATH=$C_INCLUDE_PATH:[path-to-install]/include` to expose it.
+## 1.3 PSW
+### 1.3.1 Quote Generation and Verification
+Following the [guide](https://github.com/intel/SGXDataCenterAttestationPrimitives) to install libraries for quote generation and verification, such as `libsgx_dcap_quoteverify.so` 
+
+# 2. LibOS install
+
+## 2.1 Occlum
+Following the [guide](https://github.com/occlum/occlum) to install the occlum libOS.
```
Comment (haosanzi) [inline docs/develop-env-install.md:16]: Does the default branch of occlum support enclave-cc? I remember `enclave-cc`  branch support this.

File: docs/develop-guideline.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+#### **rats-tls**
+The [rats-tls](https://github.com/inclavare-containers/inclavare-containers/tree/master/rats-tls) is a library for communicating on a secured channel, which supports the native attestation agent with feature `eaa kbc`.
+
+The rats-tls depends on the SGX SDK (following the [installing guide](develop-env-install.md)), libraries `libsgx-dcap-ql` and `libsgx-dcap-ql-dev` (install them if necessary).
+
+The rats-tls provides a different version for the application on the occlum or on the host. We want to build and run a KBC (Key Broken Service) supporting `eaa kbc` on the host, we should build and install rats-tls for the host. The library will be installed in `/usr/local/lib/rats-tls`.
+```bash
+git clone https://github.com/inclavare-containers/rats-tls.git
+cd rats-tls
+cmake -DRATS_TLS_BUILD_MODE="host" -DBUILD_SAMPLES=on -H. -Bbuild
```
Comment (haosanzi) [inline docs/develop-guideline.md:47]: please change BUILD_MODE to `occlum`

File: docs/develop-env-install.md (hunk 3 above)
Comment (HaokunX-intel) [reply docs/develop-env-install.md:16]: Sure. Should use the `enclave-cc` branch.

File: tools/debug/Makefile [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#
+# Build
+#
+
+.PHONY: build
+build:
+	cargo build --example server
+	cargo build --example client
+	cargo build --example async-server
+	cargo build --example async-client
```
Comment (haosanzi) [inline tools/debug/Makefile:10]: I think only async-client is needed

File: tools/debug/async-client.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() {
+    let matches = App::new("enclave-agent")
+                    .author("Enclave-cc Team")
+                    .arg(Arg::with_name("connect")
+                        .short("c")
+                        .long("connect")
+                        .value_name("sockaddr")
+                        .help("connect to server,  tcp://ip_addr:port")
```
Comment (haosanzi) [inline tools/debug/async-client.rs:22]: please remove the empty space in ` server,  tcp:`

**Comment by haosanzi** [comment]
I suggest splitting this pr into several commits, which are convenient for review. Thank you very much!

File: docs/develop-guideline.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+```bash
+git clone https://github.com/inclavare-containers/rats-tls.git
+cd rats-tls
+cmake -DRATS_TLS_BUILD_MODE="occlum" -DBUILD_SAMPLES=on -H. -Bbuild
+make -C build install
+```
+The library is also installed in the `/usr/local/lib/rats-tls` in default. If we want to keep two versions on the same machine, we need to install the occlum version instance into a customized place `[rats-tls-path]` and refer it in a config of the occlum instance.
+
+### *Agent Bundle Building*
+There are three steps to build a occlum-based enclave agent. Copy following three scripts under directory `enclave-cc/src` and run the command to create the bundle.
```
Comment (haosanzi) [inline docs/develop-guideline.md:61]: Please unify the markdown format.
For example: the title and the file are separated by one line

File: tools/debug/protocols/sync/mod.rs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+pub mod image;
+pub mod image_ttrpc;
+pub mod agent;
+pub mod agent_ttrpc;
+pub mod empty;
+pub mod health;
+pub mod health_ttrpc;
```
Comment (haosanzi) [inline tools/debug/protocols/sync/mod.rs:12]: I think these protocols(agent,empty,health etc) are for kata CC, they are not needed in the enclave cc scenario.

**Comment by HaokunX-intel** [comment]
> I suggest splitting this pr into several commits, which are convenient for review. Thank you very much!

This PR will only contain the debug tool and guideline will be moved to another one.

**Comment by mythi** [comment]
> > I suggest splitting this pr into several commits, which are convenient for review. Thank you very much!
> 
> This PR will only contain the debug tool and guideline will be moved to another one.

also for the future PRs, please ensure the "local development" follows/uses the tooling we already have in `tools/packaging/build`.

**Comment by HaokunX-intel** [comment]
The guideline is moved to issue #67.

**Comment by dcmiddle** [comment]
Hi @HaokunX-intel, could you describe the motivation of this PR a little more? I'm not sure what "to help query the server" means. Is this to debug access to a container image registry for example? 
Also did these tools pre-exist in another project? If so mentioning that can help ease the review.
Thanks

**Comment by HaokunX-intel** [comment]
> Hi @HaokunX-intel, could you describe the motivation of this PR a little more? 

The motivation of this PR is to provide a tool to help debug the enclave agent in the development env.  

As we known, the enclave agent is a server with ttrpc APIs. In the product env, a query such as pulling a image is passed to the enclave agent through a long path `kubelet -> containerd -> shim -> enclave agent`. In the development env, it consumes too much time to launch each component. This PR provides a little `cli` to query the enclave agent in a quick path `cli -> enclave agent`.

The tool is only used in the development env and won't be installed in the product env. And the enclave agent will be tested in the long path too.

> I'm not sure what "to help query the server" means. Is this to debug access to a container image registry for example? 

`query the server` means that the `cli` invokes the ttrpc APIs provided by the enclave agent for debugging.

> Also did these tools pre-exist in another project? If so mentioning that can help ease the review. Thanks

No. It is a debug tool created by @zhiwei-intel-h originally in local. After approved by @zhiwei-intel-h , I integrate it into this repo.

I think I need to append more details in the commit messages.

**Pull request closed without merging**